doc/design/switch_kernel.md

```diff
@@ -0,0 +1,66 @@
+## Background
+Every operator has many kernels. It because there are multiple data types, places, data layout that Fluid supports. We use the `KernelType` to describe kernel types that operators can hold. 
+
+The `KernelType` is as follows.
+
+```
+struct KernelType {
+  Place place_;
+  DataType data_type_;
+  LayoutType layout_;
+};
+```
+
+The `place_` is a descriptor of the device and the computational library, e.g., `MKLPlace`, `CUDAPlace`.
+
+The `data_type_` is the data type that this kernel performs on, e.g., `FP32`, `INT64`. Note that one kernel may have inputs with different data types. However, it will be a major `data_type`. For example, the `cross_entropy` takes `int64` as it label, and `double`/`float` as its input logit and output cost. The major `data_type` of `cross_entropy` is `float`/`double`.
+
+The `layout` is useful for some computational library. One example is that MKLDNN uses many kinds of layout, such as `nChw8`. Each kind of layout will invoke the different kernel.
+
+## Problem
+
+We register a kernel for every operator and every kernel type ideally. However, it is impracticable for the following situations.
+
+1. Some operators, like CRF, are complicated and inefficient to be implemented on GPU. The CRF operator will only have a CPU kernel.
+2. Some operators will take too many memory. It is better to force them into CPU. However, the rest of operators in this neural network will be performed in GPU, i.e., model parallel problem.
+3. Some layout and place are particular. One example is that MKLDNN uses `nChw8` and there is no other library uses `nChw8`.
+
+Problems under these situations are similar. We can formalise this problem as follow.
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
+
+## Solution
+
+It is clearly that transforming inputs of an operator from a specific kernel type to another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
+
+We can infer a kernel type from the inputs of an operators. We let this kernel type as `actual kernel type`, which means this kernel type is the actually kernel type that operator should be performed.
+
+We can get a kernel type by 1) The configuration of operator description. (Users may want to force use `MKL` for `conv` operator). 2) The place of the current executor. (Executor is running on GPU). This kernel type is what we expect the operator will be performed on. We let this kernel type as `expect kernel type`.
```

Comment:
Question: Thought data layout is determined by the actual data feed to the operator, which is the user should care. If `MKL` use different ones, let the kernel transform it just in the `MKL` kernel, will this work?

Reply:
How can we get the expected kernel type and what determines the expected type?

Reply:
So, the configuration of operator description need to be exposed to users and be set in ProgramDesc?

Reply:
> If MKL uses different ones, let the kernel transform it just in the MKL kernel, will this work?
It is not efficient.
There are many operators will be implemented by MKL.  `OpA --> OpB --> OpC`. If `OpB` and `OpC` use the MKL optimized layout format, we just transform data in `OpB`.

Reply:
> How can we get the expected kernel type and what determines the expected type?
Actually, I mean is this type really expected?

The `expected kernel type` can be determined by

1) Directly configured by users in the python program. It could be attributes of operators. e.g., There could be `force_cpu` in some operators and `use_mkldnn` in some operators. 

   We can give a unique way to configure kernel type for every kernel. It could be decided later since it does not block the work in this PR. Either an attribute named `use_mkldnn` or an unify attribute named `kernel type` is both okay.

2)

Reply:
> So, the configuration of operator description need to be exposed to users and be set in ProgramDesc?

We can give a unique way to configure kernel type for every kernel. It could be decided later since it does not block the work in this PR. Either an attribute named use_mkldnn or a unify attribute named kernel type are both okay.

Reply:
> There are many operators will be implemented by MKL. OpA --> OpB --> OpC. If OpB and OpC use the MKL optimized layout format, we just transform data in OpB.


"DataLayout" is one attribute of the typed data (or LoDTensor which also have lod as its attribute), in the [Dataflow](https://en.wikipedia.org/wiki/Dataflow_programming) model, we prefer less states of the kernel. For efficiency, kernels do not need to transform the layout if it is already satisfied.